https://bugs.gentoo.org/913339
--- a/gcc/common.opt
+++ b/gcc/common.opt
@@ -3206,7 +3206,7 @@ Common Var(flag_tree_scev_cprop) Init(1) Optimization
 Enable copy propagation of scalar-evolution information.
 
 ftrivial-auto-var-init=
-Common Joined RejectNegative Enum(auto_init_type) Var(flag_auto_var_init) Init(AUTO_INIT_UNINITIALIZED) Optimization
+Common Joined RejectNegative Enum(auto_init_type) Var(flag_auto_var_init) Init(AUTO_INIT_ZERO) Optimization
 -ftrivial-auto-var-init=[uninitialized|pattern|zero]	Add initializations to automatic variables.
 
 Enum